### DISCUSSION TITLE
How to best update multi-value properties?

### DISCUSSION THREAD

**Discussion by dpriskorn** [LeMyst/WikibaseIntegrator#399] [Q&A] [answered]
Hi. I want to update values of a property to reflect reality. 
E.g. wcdimportbot reads the Apollo 1 article in enwiki and finds 10 references on the first pass. A month later one has been removed and one new has been added. On a second pass I want to delete the removed one and add the new one.

Right now I let WBI do all the work by simply having this line which only adds new references:
https://github.com/internetarchive/wcdimportbot/blob/feb0d7b8c40467b0420c8a5d3b0be380efa8c962/src/models/wikibase/crud/update.py#L102
That does not remove any references that no longer appear in the article from the property.

I'm trying to devise a way to remove values. Would this work:
1) iterate over all values in the property and remove() them if not found in the new list of values from the article.
2) add new claims like I do now.

**Comment by dpriskorn** [comment, marked as answer]
The strategy above seems to work fine. It can now both delete and add values as expected 😀
See https://sandbox.wiki/w/index.php?title=Item%3AQ7437&type=revision&diff=4315&oldid=4314

**Reply by LeMyst** [reply]
This changes probably helped a lot:
https://github.com/internetarchive/wcdimportbot/commit/64867f7a1ddc7910bfafb8b335e03eeb1d2c2a6b#diff-74f6ee523bdc2170aa76e8fc9f72cf31636fb1b20756be189a30044a8f615d28R101

**Comment by LeMyst** [comment]
Hello Dennis,

I'm afraid to not fully understand what you want to do. But your goal seems to replace all your claims (and references) without keeping the existing?
Firstly, you seems to use a mix between `self.updated_claims` and `updated_claims`, I don't know if that can changes the result wanted.
Otherwise, maybe you can create a list a `new_claims` and outside the loop update the claims with on `updated_claims.add(claims=new_claims_list)`

**Reply by dpriskorn** [reply]
Yes, that was a bug to mix the two. I fixed that 😀
